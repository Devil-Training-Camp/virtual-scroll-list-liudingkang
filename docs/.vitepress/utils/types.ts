import { DefaultTheme, LocaleSpecificConfig } from 'vitepress';
export type Locales = LocaleSpecificConfig<DefaultTheme.Config> & { label: string; link?: string };
