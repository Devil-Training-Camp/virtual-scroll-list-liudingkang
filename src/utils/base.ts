export interface AnyObject {
  [key: number | string]: unknown;
}
