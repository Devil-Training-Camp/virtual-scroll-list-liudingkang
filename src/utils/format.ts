export function camelize(str: string): string {
  return str.toLocaleLowerCase();
}
