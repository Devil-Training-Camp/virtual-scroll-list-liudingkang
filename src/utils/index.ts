export * from './base';
export * from './with-intall';
export * from './format';
export * from './optimize';
