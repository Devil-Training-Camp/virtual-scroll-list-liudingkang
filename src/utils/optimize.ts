type Timer = null | number;
export type RestFunc = <T>(...args: T[]) => void;
export function throttle<T extends RestFunc>(func: T, time = 0, immediate = true) {
  let timer: Timer = null;
  return function (this: unknown, ...args: Parameters<T>) {
    if (immediate) {
      func.call(this, ...args);
      immediate = false;
    }
    if (timer == null) {
      timer = setTimeout(() => {
        func.call(this, ...args);
        timer = null;
      }, time);
    }
  };
}
export function RAFThrottle<T extends RestFunc>(func: T) {
  let flag = false;
  return function (this: unknown, ...args: Parameters<T>) {
    if (!flag) {
      window.requestAnimationFrame(() => {
        func.call(this, ...args);
        flag = false;
      });
      flag = true;
    }
  };
}
export const BSStartIndex = (arr: number[], num: number) => {
  let left = 0,
    right = arr.length - 1;

  while (left != right) {
    const mid = left + ((right - left) >> 1);
    if (arr[mid] > num) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return Math.max(0, left - 1);
};
