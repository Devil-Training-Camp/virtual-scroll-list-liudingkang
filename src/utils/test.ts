import { nextTick } from 'vue';

export const delay = (time?: number) => {
  return new Promise(resolve => {
    setTimeout(resolve, time);
  });
};
export const trigger = async (target: HTMLElement, type: string) => {
  const scrollEvent = new Event(type, {
    bubbles: true,
    cancelable: true,
  });
  target.dispatchEvent(scrollEvent);
  await nextTick();
};
export const triggerScrollTo = async (target: HTMLElement, x: number = 0, y: number = 0) => {
  target.scrollTo({
    left: x,
    top: y,
  });
  await trigger(target, 'scroll');
};
