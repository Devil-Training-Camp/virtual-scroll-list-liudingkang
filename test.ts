interface VInfo {
  major: string;
  minor: string;
  patch: string;
  prerelease: string;
  preItems: string[];
  compileInfo: string;
}

function versionCompare(v1: string, v2: string): number {
  const p1 = parseV(v1);
  const p2 = parseV(v2);
  switch (finalCompare(p1, p2)) {
    case 1:
      return 1;
    case -1:
      return -1;
  }
  preCompare(p1, p2);
}
function finalCompare(p1: VInfo, p2: VInfo): number {
  const { major: major1, minor: minor1, patch: patch1 } = p1;
  const { major: major2, minor: minor2, patch: patch2 } = p2;
  switch (strNumCompare(major1, major2)) {
    case 1:
      return 1;
    case -1:
      return -1;
  }
  switch (strNumCompare(minor1, minor2)) {
    case 1:
      return 1;
    case -1:
      return -1;
  }
  switch (strNumCompare(patch1, patch2)) {
    case 1:
      return 1;
    case -1:
      return -1;
  }
  return 0;
}
function preCompare(p1: VInfo, p2: VInfo): number {
  const { preItems: preItems1 } = p1;
  const { preItems: preItems2 } = p2;
  const maxLen = Math.max(preItems1.length, preItems2.length);
  for (let i = 0; i < maxLen; i++) {
    const item1 = preItems1[i];
    const item2 = preItems2[i];
    // 有一个不存在
    if (preItems1[i].length === 0 || preItems2[i].length === 0) {
      return preItems1[i].length > 0 ? 1 : -1;
    }
    // 两个都存在
    if (isStrNum(item1) && isStrNum(item2)) {
      // 都由数字组成，比较数字大小
      switch (strNumCompare(item1, item2)) {
        case 1:
          return 1;
        case -1:
          return -1;
      }
    } else {
      // 有一个不为数字，按 ASCII 顺序比较
    }
  }
  return 0;
}
function strNumCompare(v1: string, v2: string): number {
  if (v1.length !== v2.length) {
    return v1.length > v2.length ? 1 : -1;
  }
  for (let i = 0, len = v1.length; i < len; i++) {
    if (v1[i] !== v2[i]) {
      return v1[i] > v2[i] ? 1 : -1;
    }
  }
  return 0;
}
function isStrNum(s: string): boolean {
  return s.length > 0 && !Number.isNaN(Number(s));
}

function parseV(v: string): VInfo {
  const reg = /(\d+?)\.(\d+?)\.(\d+?)-?([\w.]*)\+?([\w.]*)/;
  const matchInfo = v.match(reg);
  return {
    major: matchInfo[1],
    minor: matchInfo[2],
    patch: matchInfo[3],
    prerelease: matchInfo[4],
    preItems: matchInfo[4].split('.'),
    compileInfo: matchInfo[5],
  };
}

console.log(versionCompare('1.2.3', '9.8.7'));
