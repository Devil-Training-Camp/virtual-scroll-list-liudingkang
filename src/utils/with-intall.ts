import type { App, Component } from 'vue';
import { camelize } from './format';

export type WithInstall<T> = T & {
  install(app: App): void;
};
// 组件 install 方法
export function withInstall<T extends Component>(options: T): WithInstall<T> {
  (options as WithInstall<T>).install = (app: App) => {
    const { name } = options;
    if (!name) return;
    app.component(name, options);
    app.component(camelize(name), options);
  };
  return options as WithInstall<T>;
}
