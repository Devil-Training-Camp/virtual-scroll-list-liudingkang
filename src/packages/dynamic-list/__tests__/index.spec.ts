import { describe, expect, test, vi } from 'vitest';
import { type Component, nextTick } from 'vue';

import { DynamicList } from '..';
import { delay, prevMount, triggerScrollTo } from '../../../utils/test';

const TestDynamicList: Component = {
  template: `
  <DynamicList v-bind="$attrs">
    <template #default="{ item, index }">
      <div
        class="test-item"
        :style="{
          height: item + 'px',
          boxSizing: 'border-box',
        }"
      >
        {{ index }}
      </div>
    </template>
  </DynamicList>`,
  components: {
    DynamicList,
  },
};

const mount = prevMount(TestDynamicList, DynamicList);
const mockData = (num = 20, height = 50) => {
  const data = [];
  for (let index = 0; index < num; index++) {
    data.push(Math.ceil((Math.random() + 0.5) * height));
  }
  return data;
};
describe('test dynamic-list component props', () => {
  const data = mockData(100);
  test('test dynamic-list data 0', async () => {
    const wrapper = mount({
      props: {
        itemSize: 50,
        data: [],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test('test dynamic-list itemSize', async () => {
    const wrapper = mount({
      props: {
        itemSize: 50,
        data,
      },
    });
    expect((wrapper.find('.test-item').element as HTMLElement).style.height).toBe(`${data[0]}px`);
    wrapper.unmount();
  });
  test('test dynamic-list itemClass', async () => {
    const wrapper = mount({
      props: {
        data,
        itemClass: 'custom-item',
      },
    });
    expect(wrapper.find('.virtual-list-item').classes()).toContain('custom-item');
  });
  test('test dynamic-list width and height', () => {
    const wrapper = mount({
      props: {
        data,
        width: 300,
        height: 600,
      },
    });
    const elm = wrapper.find('.virtual-list-container').element as HTMLElement;
    expect(elm.style.width).toBe('300px');
    expect(elm.style.height).toBe('600px');
  });
  test('test dynamic-list cache', () => {
    const wrapper = mount({
      props: {
        data,
        height: 600,
        itemSize: 155,
        cache: 2,
      },
    });
    expect(wrapper.findAll('.virtual-list-item')).toHaveLength(6);
  });
});
describe('test dynamic-list scroll', () => {
  test('test dynamic-list scroll cache', async () => {
    const data = mockData(100, 150);
    const wrapper = mount({
      props: {
        data,
        height: 600,
        itemSize: 150,
        cache: 4,
      },
    });
    expect(wrapper.findAll('.virtual-list-item').length).toBeGreaterThan(6);
    expect(wrapper.findAll('.virtual-list-item').length).toBeLessThan(10);
    const elm = wrapper.find('.virtual-list-container').element as HTMLElement;
    await triggerScrollTo(elm, 0, 650);
    await delay();
    expect(wrapper.findAll('.virtual-list-item').length).toBeGreaterThan(10);
    expect(wrapper.findAll('.virtual-list-item').length).toBeLessThan(14);
    await triggerScrollTo(elm, 0, 1650);
    await delay();
    expect(wrapper.findAll('.virtual-list-item').length).toBeGreaterThan(10);
    expect(wrapper.findAll('.virtual-list-item').length).toBeLessThan(14);
  });
  test('test dynamic-list scroll distance', async () => {
    const data = mockData(5, 150);
    const onLoad = vi.fn();

    const wrapper = mount({
      props: {
        data,
        height: 600,
        itemSize: 150,
        distance: 100,
        onLoad,
      },
    });
    const elm = wrapper.find('.virtual-list-container').element as HTMLElement;
    await triggerScrollTo(elm, 0, data.reduce((prev, item) => (prev += item), 0) - 100);
    await delay();
    expect(onLoad).toBeCalledTimes(1);
  });
});
describe('test dynamic-list event', () => {
  const data = mockData(5, 150);
  test('test dynamic-list onload', async () => {
    const onLoad = vi.fn();
    const wrapper = mount({
      props: {
        data,
        height: 600,
        itemSize: 150,
        onLoad,
      },
    });
    const elm = wrapper.find('.virtual-list-container').element as HTMLElement;
    await triggerScrollTo(elm, 0, 150);
    await delay();
    expect(onLoad).toBeCalledTimes(1);
  });
});

describe('test dynamic-list item', () => {
  const data = mockData(5, 150);
  test('test dynamic-list item resize', async () => {
    const onLoad = vi.fn();
    const wrapper = mount({
      props: {
        data,
        height: 600,
        itemSize: 150,
        onLoad,
      },
    });
    const elm = wrapper.find('.test-item').element as HTMLElement;
    // 直接修改元素高度并不能触发 resizeObserver 回调，可能是 happy-dom 的局限性，在 chrome 上能正常触发
    elm.style.height = '50px';
    await nextTick();
    expect(elm.style.height).toBe(`50px`);
  });
});
