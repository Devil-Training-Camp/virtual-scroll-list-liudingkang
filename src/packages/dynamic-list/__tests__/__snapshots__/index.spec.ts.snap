// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test dynamic-list component props > test dynamic-list data 0 1`] = `
"<div data-v-9610e607="" class="virtual-list-container dynamic-list" style="width: 100%; height: 300px;">
  <div data-v-9610e607="" class="virtual-list" style="width: 100%; height: 0px;"></div>
</div>"
`;
